lib/spack/spack/environment.py

```diff
@@ -928,9 +942,11 @@ def _to_lockfile_dict(self):
         concrete_specs = {}
         for spec in self.specs_by_hash.values():
             for s in spec.traverse():
-                dag_hash = s.dag_hash()
-                if dag_hash not in concrete_specs:
-                    concrete_specs[dag_hash] = s.to_node_dict(all_deps=True)
+                dag_hash_all = s.dag_hash(all_deps=True)
```

Comment:
It probably is my fault since I either wrote these variables in or requested that they be added in review. IMO it would be more confusing to add `deptype` in because these functions don't have the same customization requirements that other `Spec` methods do (would it make sense to ask for a DAG hash following only the link dependencies? or only the run dependencies?).

I could be convinced otherwise but I would definitely want to take you up on the offer to deal with it in another PR.